prow/cmd/build/controller.go

```diff
@@ -17,46 +17,424 @@ limitations under the License.
+	return c.bLister.Builds(namespace).Get(name)
+}
+func (c *controller) deleteBuild(namespace, name string) error {
+	return c.bc.BuildV1alpha1().Builds(namespace).Delete(name, &metav1.DeleteOptions{})
+}
+func (c *controller) createBuild(namespace string, b *buildv1alpha1.Build) (*buildv1alpha1.Build, error) {
+	return c.bc.BuildV1alpha1().Builds(namespace).Create(b)
+}
+func (c *controller) updateBuild(namespace string, b *buildv1alpha1.Build) (*buildv1alpha1.Build, error) {
+	return c.bc.BuildV1alpha1().Builds(namespace).Update(b)
+}
+func (c *controller) updateProwJob(namespace string, pj *prowjobv1.ProwJob) (*prowjobv1.ProwJob, error) {
+	return c.pjc.ProwV1().ProwJobs(namespace).Update(pj)
+}
+
+func (c *controller) now() metav1.Time {
+	return metav1.Now()
+}
+
+var (
+	groupVersionKind = schema.GroupVersionKind{
+		Group:   prowjobv1.SchemeGroupVersion.Group,
+		Version: prowjobv1.SchemeGroupVersion.Version,
+		Kind:    "ProwJob",
+	}
+)
+
+// reconcile ensures that the prowjob and build
+func reconcile(c reconciler, key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("invalid resource %s: %v", key, err))
+	}
+
+	var wantBuild bool
+
+	pj, err := c.getProwJob(namespace, name)
+	switch {
+	case apierrors.IsNotFound(err):
+		// Do not want build
+	case err != nil:
+		return fmt.Errorf("get prowjob: %v", err)
+	case pj.Spec.Agent != prowjobv1.BuildAgent:
+		// Do not want a build for this job
+	case pj.DeletionTimestamp == nil:
+		wantBuild = true
+	}
+
+	var haveBuild bool
+
+	b, err := c.getBuild(namespace, name)
+	switch {
+	case apierrors.IsNotFound(err):
+		// Do not have a build
+	case err != nil:
+		return fmt.Errorf("get build: %v", err)
+	case b.DeletionTimestamp == nil:
+		haveBuild = true
+	}
+
+	// Should we create or delete this build?
+	switch {
+	case !wantBuild:
+		if !haveBuild {
+			logrus.Infof("Deleted %s", key)
+			return nil
+		}
+		switch or := metav1.GetControllerOf(b); {
+		case or == nil, or.APIVersion != groupVersionKind.GroupVersion().String(), or.Kind != groupVersionKind.Kind:
+			return nil // Not controlled by this
+		}
+		logrus.Infof("Delete builds/%s", key)
+		if err = c.deleteBuild(namespace, name); err != nil {
+			return fmt.Errorf("delete build: %v", err)
+		}
+		return nil
+	case finalState(pj.Status.State):
+		logrus.Infof("Finished %s", key)
+		return nil
+	case wantBuild && pj.Spec.BuildSpec == nil:
+		return errors.New("nil BuildSpec")
+	case wantBuild && !haveBuild:
+		if b, err = makeBuild(*pj); err != nil {
+			return fmt.Errorf("make build: %v", err)
+		}
+		logrus.Infof("Create builds/%s", key)
+		if b, err = c.createBuild(namespace, b); err != nil {
+			return fmt.Errorf("create build: %v", err)
+		}
+	}
+
+	// Ensure spec is correct
+	pj.Spec.BuildSpec.Generation = b.Spec.Generation
+	if !reflect.DeepEqual(b.Spec, *pj.Spec.BuildSpec) {
```

Comment:
`equality.Semantic.DeepEqual()` for comparing objects with semantic meanings that are not generic to data (lists without regard to ordering, etc)

Reply:
Done